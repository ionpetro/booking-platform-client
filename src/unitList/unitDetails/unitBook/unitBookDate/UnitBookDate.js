import Input from '../../../../components/Input/Input.vue';

export default {
  components: {
    'v-input': Input
  },
  data: () => ({
    show: false,
    selectedYear: null,
    years: ['2080', '2081', '2082', '2083', '2084', '2085', '2086', '2087', '2088']
  }),
  props: {
    unit: {
      required: true
    }
  },
  methods: {
    disabledYears(year) {
      return this.unit.reservedYears.includes(+year);
    },
    hideModal() {
      this.show = false;
    },
    toggleModal() {
      this.show = !this.show;
    },
    selectYear(year) {
      if (this.disabledYears(year)) {
        return;
      }
      this.selectedYear = year;
      this.$emit('onSelectedYear', year);
    }
  }
};
