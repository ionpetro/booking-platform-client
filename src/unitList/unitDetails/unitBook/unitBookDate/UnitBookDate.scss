.bookDate {
  input {
    cursor: pointer;
  }

  &__modal {
    display: grid;
    grid-template-columns: auto auto auto;
    padding: 10px;

    position: absolute;
    width: 179px;
    height: 136px;
    background: white;
    border-radius: 4px;
    bottom: 100px;
    filter: drop-shadow(0px 4px 15px rgba(0, 0, 0, 0.2));

    &__year {
      cursor: pointer;
      display: flex;
      justify-content: center;
      align-items: center;
      color: $gray-dark;

      &:hover {
        background: $gray-light;
        border-radius: 2px;
      }

      &--disabled {
        cursor: not-allowed;
        background: $gray-tint;
        color: $gray-dim;
        &:hover {
          background: $gray-tint;
        }
      }

      &--selected {
        background-color: $secondary;
        color: white;
        &:hover {
          background: $secondary;
        }
      }
    }
  }

  &__input {
    position: relative;
    &__calendar {
      position: absolute;
      left: 8px;
      bottom: 20px;
    }
    &__reset {
      cursor: pointer;
      position: absolute;
      right: 20px;
      bottom: 10px;
      padding: 15px 4px;
      transition: transform 0.3s ease-in-out;
      transform: rotate(180deg);
    }
    &__reverse {
      cursor: pointer;
      position: absolute;
      right: 20px;
      bottom: 10px;
      padding: 15px 4px;
      transition: transform 0.3s ease-in-out;
      transform: rotate(0deg);
    }
    &__text {
      border: 1px solid $secondary;
      box-sizing: border-box;
      height: 56px;
      width: 103px;
      padding-left: 32px;
      border-radius: 10px;
      color: $secondary;
      font-size: 1rem;
      margin-right: 8px;

      &:focus {
        outline: none;
      }
    }
  }

  &__label {
    color: $secondary;
    position: absolute;
    z-index: 1;
    left: 8px;
    top: -10px;
    padding: 0 8px;
    background-color: white;
  }
}
.fade-leave-active {
  transition: opacity 0.5s;
}
.fade-leave-to {
  opacity: 0;
}
