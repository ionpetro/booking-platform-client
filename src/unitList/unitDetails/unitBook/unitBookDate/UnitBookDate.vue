<template>
  <div class="bookDate" v-click-outside="hideModal">
    <div class="bookDate__input">
      <img class="bookDate__input__calendar" src="@/assets/icons/Calendar.svg" alt="Calendar" />
      <img
        :class="['testArrow', show ? 'bookDate__input__reset' : 'bookDate__input__reverse']"
        src="@/assets/icons/Arrow.svg"
        alt="Arrow"
        @click="toggleModal"
      />
      <input
        class="bookDate__input__text"
        :value="selectedYear"
        placeholder="YYYY"
        @click.prevent="toggleModal"
        readonly
      />
      <label class="bookDate__label body--small" for="">Year</label>
    </div>
    <transition name="fade">
      <div v-if="show">
        <div class="bookDate__modal">
          <span
            v-for="(year, index) of years"
            :key="`year${index}`"
            @click="selectYear(year)"
            :class="[
              'bookDate__modal__year',
              disabledYears(year) ? 'bookDate__modal__year--disabled' : '',
              year === selectedYear ? 'bookDate__modal__year--selected' : ''
            ]"
            tabindex="0"
            >{{ year }}</span
          >
        </div>
      </div>
    </transition>
  </div>
</template>

<script src="./UnitBookDate.js"></script>

<style lang="scss" src="./UnitBookDate.scss" scoped></style>
