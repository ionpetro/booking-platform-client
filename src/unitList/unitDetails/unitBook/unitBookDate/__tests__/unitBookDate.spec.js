import {mount} from '@vue/test-utils';
import UnitBookDate from '../UnitBookDate.vue';
import mockUnit from '../../../__tests__/mocks/unit.mock';

describe('UnitBookDate', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = mount(UnitBookDate, {
      propsData: {
        unit: mockUnit
      }
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  describe('when loaded', () => {
    it('renders data correctly', async () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('when input arrow is clicked', () => {
    it('opens modal', async () => {
      // click arrow button
      await wrapper.find('.testArrow').trigger('click');

      // expect the modal to be open
      expect(wrapper.vm.$data.show).toBeTruthy();
    });

    describe('when modal is open', () => {
      it('selectes year when it is clicked', async () => {
        // wrapper.vm.$data.show = true;
        await wrapper.find('.testArrow').trigger('click');

        // year 2080 is clicked (it is enabled)
        await wrapper
          .findAll('span')
          .at(2)
          .trigger('click');

        expect(wrapper.emitted().onSelectedYear[0]).toEqual(['2082']);
      });
      it('does nothing when disabled year is clicked', async () => {
        // wrapper.vm.$data.show = true;
        await wrapper.find('.testArrow').trigger('click');

        // year 2081 is clicked (it is disabled)
        await wrapper
          .findAll('span')
          .at(1)
          .trigger('click');

        expect(wrapper.emitted().onSelectedYear).toBeFalsy();
      });
    });
  });
});
