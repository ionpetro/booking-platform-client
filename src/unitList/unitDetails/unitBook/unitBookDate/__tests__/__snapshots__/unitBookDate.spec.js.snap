// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitBookDate when loaded renders data correctly 1`] = `
<div
  class="bookDate"
>
  <div
    class="bookDate__input"
  >
    <img
      alt="Calendar"
      class="bookDate__input__calendar"
      src="@/assets/icons/Calendar.svg"
    />
     
    <img
      alt="Arrow"
      class="testArrow bookDate__input__reverse"
      src="@/assets/icons/Arrow.svg"
    />
     
    <input
      class="bookDate__input__text"
      placeholder="yyyy"
      readonly="readonly"
    />
     
    <label
      class="bookDate__label body--small"
      for=""
    >
      Year
    </label>
  </div>
   
  <transition-stub
    name="fade"
  >
    <!---->
  </transition-stub>
</div>
`;
